### PR TITLE
Feature merge2

### PR DESCRIPTION
**Pull request by chryswoods**
This is the finished `feature_merge` pull request, that brings in all of the functionality needed for `2024.1.0`, namely support for creating merge molecules, decoupling, annihilation, and better debugging of lambda schedules. I've also added in pow support for GeneralUnit.

I've added all the tests, plus now fully describe the new functionality in a new part 7 of the tutorial.

Once merged, I'll get everything ready for the `2024.1.0` release :-)

* I confirm that I have merged the latest version of `devel` into this branch before issuing this pull request (e.g. by running `git pull origin devel`): [y]
* I confirm that I have added a test for any new functionality in this pull request: [y]
* I confirm that I have added documentation (e.g. a new tutorial page or detailed guide) for any new functionality in this pull request: [y]
* I confirm that I have added a changelog entry to the changelog (we will add a link to this PR as part of the review): [y/n]
* I confirm that I have permission to release this code under the GPL3 license: [y]

## Suggested reviewers:
@lohedges

## Any additional context of information?

I've run all the BioSimSpace tests, plus sire_bigtests, plus checked that the ethane/methanol free energy is still correct. Not much has changed functionally since feature_merge - just some additional functionality as described in the tutorial (and also experimented with in Julien's lab).

### CODE REVIEW

**Comment by chryswoods** [comment]
Weird conda build issues - will see what happens tomorrow.

**Comment by chryswoods** [comment]
The Windows failure is a new gemmi wrapping issue:

```
2024-03-31T01:33:41.6639422Z     gemmi:                     0.6.5-py311h28e1466_0          conda-forge
2024-03-31T01:36:08.4980148Z tests/convert/test_gemmi.py::test_gemmi FAILED                           [  2%]
2024-03-31T01:36:08.4980998Z tests/convert/test_gemmi.py::test_gemmi_roundtrip FAILED                 [  2%]
2024-03-31T01:36:08.4981894Z tests/convert/test_gemmi.py::test_gemmi_complex_metadata PASSED          [  2%]
2024-03-31T01:36:19.4580529Z _________________________________ test_gemmi __________________________________
2024-03-31T01:36:19.4585981Z         "gemmi" not in sr.convert.supported_formats(), reason="gemmi not available"
2024-03-31T01:36:19.4588308Z     def test_gemmi(testfile_cache_dir, pdbx_3nss):
2024-03-31T01:36:19.4593506Z         import gemmi
2024-03-31T01:36:19.4594765Z         s = gemmi.read_structure(cif_file)
2024-03-31T01:36:19.4597112Z tests\convert\test_gemmi.py:26:
2024-03-31T01:36:19.4602237Z     c = gemmi_to_sire(typ[1], map=map)
2024-03-31T01:36:19.4603964Z obj = [<gemmi.Structure 3NSS with 1 model(s)>], map = {}
2024-03-31T01:36:19.4605101Z     def gemmi_to_sire(obj, map=None):
2024-03-31T01:36:19.4606999Z         Convert the passed gemmi Structure to the sire equivalent
2024-03-31T01:36:19.4610054Z             from ..legacy.Convert import gemmi_to_sire as _gemmi_to_sire
2024-03-31T01:36:19.4614064Z                 "gemmi is not available. Please install via "
2024-03-31T01:36:19.4615661Z                 "'conda install -c conda-forge gemmi'"
2024-03-31T01:36:20.1158114Z >           s = System(_gemmi_to_sire(o, map=map))
2024-03-31T01:36:20.1159596Z E               sire.legacy.Convert._SireGemmi.gemmi_to_sire(gemmi.Structure)
2024-03-31T01:36:20.1161174Z E               gemmi_to_sire(struct gemmi::Structure mols, class SireBase::PropertyMap map)
2024-03-31T01:36:20.1162775Z ____________________________ test_gemmi_roundtrip _____________________________
2024-03-31T01:36:20.1163729Z tmpdir = local('C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pytest-of-runneradmin\\pytest-0\\test_gemmi_roundtrip0')
2024-03-31T01:36:20.1166439Z         "gemmi" not in sr.convert.supported_formats(), reason="gemmi not available"
2024-03-31T01:36:20.1168115Z     def test_gemmi_roundtrip(tmpdir, pdbx_3nss):
2024-03-31T01:36:20.1179421Z >       g = sr.convert.to(mols, "gemmi")
2024-03-31T01:36:20.1180000Z tests\convert\test_gemmi.py:64:
2024-03-31T01:36:20.1181755Z     return to_gemmi(obj, map=map)
2024-03-31T01:36:20.1182492Z ..\_test_env\Lib\site-packages\sire\convert\__init__.py:200: in to_gemmi
2024-03-31T01:36:20.1183445Z     return sire_to_gemmi(to_sire(obj, map=map), map=map)
2024-03-31T01:36:20.1185722Z     def sire_to_gemmi(obj, map=None):
2024-03-31T01:36:20.1186829Z         Convert the passed sire object to a gemmi structure
2024-03-31T01:36:20.1188707Z             from ..legacy.Convert import sire_to_gemmi as _sire_to_gemmi
2024-03-31T01:36:20.1206446Z >           g = _sire_to_gemmi(o, map=map)
2024-03-31T01:36:20.1207114Z E           TypeError: Unregistered type : gemmi::Structure
2024-03-31T01:36:20.1222746Z FAILED tests/convert/test_gemmi.py::test_gemmi - Boost.Python.ArgumentError: Python argument types in
2024-03-31T01:36:20.1223975Z     sire.legacy.Convert._SireGemmi.gemmi_to_sire(gemmi.Structure)
2024-03-31T01:36:20.1224978Z     gemmi_to_sire(struct gemmi::Structure mols, class SireBase::PropertyMap map)
2024-03-31T01:36:20.1225988Z FAILED tests/convert/test_gemmi.py::test_gemmi_roundtrip - TypeError: Unregistered type : gemmi::Structure
```

**Comment by lohedges** [comment]
This is because of [this] (https://github.com/conda-forge/boa-feedstock/issues/81) issue that I reported after one of my recent BioSimSpace PRs failed. They added a fix, but it didn't solve the second issue. Pinning to `boa` version 0.16 fixes things for now.

I'm away this week but will hopefully find time to review.

**Comment by chryswoods** [comment]
Thanks - I'll add back in mamba and will pin boa.

The Windows gemmi issue is strange. It all works locally (even using the newer gemmi package). I've resubmitted the Windows build to check for weirdness. My guess is that it is some change brought about somewhere between gemmi and pybind11, or with the compiler.

**Comment by mbargull** [comment]
Hi, I saw this via the https://github.com/conda-forge/boa-feedstock/issues/81 back reference.
The reason you encountered the `boa`/`conda`/`boltons` issue here stems from the `packaging=21` pin in your workflow files.
This `packaging=21` constraint prevents newer `conda` versions from being installed since `conda>=23.3.0` has a dependency on `packaging>=23.0`

**Comment by lohedges** [comment]
Many thanks. We can remove the `packaging` pin and see what happens. Presumably that was added to work around a previous issue, so hopefully it's no longer needed.

**Comment by chryswoods** [comment]
Thanks - I'll try this now :-)

**Comment by chryswoods** [comment]
Thanks - I confirm that removing the `packaging=21` has fixed the boa issue - everything is now compiling on our supported platforms.

Lester - I missed a gemmi pin for the Windows build. I've added this back in as ci-skip, as I want to see if the current build fixes the problem. If not, I will re-run it. The pin change doesn't impact Mac or Linux, so no need to re-run those.

**Comment by lohedges** [comment]
Great, thanks for confirming. I'll update the BioSimSpace builds when I do the release there.

**Comment by chryswoods** [comment]
Updating more for my own record...

Windows build with both gemmi 0.6.5 and gemmi 0.6.4 failed for the same reason, so it is not any change with that package. I am trying harder to see if I can recreate this locally, as fixing this on GH will be too painful...

**Comment by chryswoods** [comment]
I can now recreate this locally, which helps. This rules out a compiler change or compile flags change. Gemmi version is the same, so maybe something changing in boost::python or pybind11. I'll keep investigating.

As a side-point, the manual install in Python 3.11 upgraded to 3.12 automatically when installing all BSS dependencies on Windows. This seems to run well. I think I will move to 3.10-3.12 for 2024.1.0. Our 3.9 packages aren't used as much - the most popular are the 3.10.

**Comment by chryswoods** [comment]
It's a problem with pybind11. Everything works using pybind11 version 2.11.1. It breaks with 2.12.

Adding some debugging shows that pybind11 is unable to detect the type of the C++ class.

```
Unable to cast Python instance of type <class 'gemmi.Structure'> to C++ type '?' (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)
```

I'll see if I can fix this. If not, it will be pinned.

**Comment by chryswoods** [comment]
I gave up fixing pybind11 on Windows. I tracked it down to code in `pybind11/cast.h`, but casting looks very complex, and not worth the energy now. The new code came out a week or so ago, so I am 🤞 that someone else will notice the break and be motivated enough to fix it ;-). It's only a build dependency, so pinning to 2.11.1 on Windows doesn't impact our runtime dependencies.

In terms of moving to Python 3.12, there's some missing dependencies for MacOS X86, but since this platform is little used (and Intel Macs are no longer sold), I don't want to have it hold up us supporting 3.12. I've thus decided to move to 3.10-3.12, with a caveat that MacOS 3.12 is pending for a point release (and detailed this in the changelog). 

I think it is good to make this change now, as it's now ~6 months since the 3.12 release and it is widely supported. 3.9 is over 3 years old and is only getting security patches.

Also, Python seems in a pattern now of making their major releases in October. We can get in the pattern of changing our supported Python versions in our first release of a year, so that this can become more predictable too.

**Comment by chryswoods** [comment]
For my notes, the specific dependencies on MacOS ARM/X86 that we are awaiting are mdtraj and openmmtools (which is blocked by missing mdtraj). There's 3.12 packages for mdtraj on other platforms, so I expect it is just a matter of time for MacOS support.

**Comment by lohedges** [comment]
Thanks for the info. `mdtraj` shouldn't really be a hard dependency anyway as it's just an optional trajectory backend which, unlike `mdanalysis`, I don't think other packages depend on. Hopefully the dependency resolution doesn't change too much when it is included.

**Comment by chryswoods** [comment]
I've also implemented the `auto-bonds` constraint as requested in wishlist item #185.

This is now finished - nothing else will go into 2024.1.0 ;-)

**Comment by chryswoods** [comment]
Windows build failed because it couldn’t install kcombu_bss - I guess a timeout or conda issue. Will rerun once the MacOS build has finished.

**Review by lohedges — APPROVED** [review]
Many thanks, this is fantastic. I really like how it's now possible to get so much information about the underlying perturbation. This will make it much easier for the user to understand what's going on, as well as massively simplifying debugging.

It's a real shame the GitHub's `reStructuredText` rendering is so broken at the moment since it makes it really hard to follow the tutorials in the repo. (The website will be fine.) The only thing that I spotted while reading through is an inconsistency in the option used to allow matching of light atoms in a mapping. 

In [this](https://github.com/OpenBioSim/sire/blob/feature_merge2/doc/source/tutorial/part07/04_merge.rst) part of the tutorial it says:

> By default, this algorithm ignores hydrogens. This is why only a single atom was matched above. You can change this by passing in the `ignore_light_atoms=True` argument.

In a [later section](https://github.com/OpenBioSim/sire/blob/feature_merge2/doc/source/tutorial/part07/07_residue.rst) you use:

```python
mapping = sr.morph.match(ala, lys, match_light_atoms=True) 
```

I assume that this is meant to be `match_light_atoms` in both cases.

**Comment by lohedges** [comment]
Everything still runs as expected, only I now see the warning above. This is definitely something that Exscientia would see. I'll try to figure out what bit of code is triggering it. I assume this is when the system is squashed, i.e. the two end states and put one after the other in another system.

**Comment by lohedges** [comment]
Yes, that was it. Running the following:

```
python -m pytest -svvv --color=yes tests/Sandpit/Exscientia/Align/test_squash.py 
```

Gives:

```python
============================================================ test session starts ============================================================
platform linux -- Python 3.11.7, pytest-7.4.4, pluggy-1.3.0 -- /home/lester/.conda/envs/openbiosim/bin/python
cachedir: .pytest_cache
rootdir: /home/lester/Code/openbiosim/biosimspace
configfile: pyproject.toml
plugins: anyio-4.2.0
collected 28 items

tests/Sandpit/Exscientia/Align/test_squash.py::test_squash[False-expected_n_atoms0]
────────────────────────────────────────────────────────────────── WARNING ──────────────────────────────────────────────────────────────────
Could not convert the old property at key parameters with type SireMM::AmberParams to match the new, edited molecule. This property has been
deleted.
Error was SireError::incomplete_code: Cannot make compatible if atom order has changed!

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

────────────────────────────────────────────────────────────────── WARNING ──────────────────────────────────────────────────────────────────
Could not convert the old property at key parameters with type SireMM::AmberParams to match the new, edited molecule. This property has been
deleted.
Error was SireError::incomplete_code: Cannot make compatible if atom order has changed!

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squash[True-expected_n_atoms1] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squash_multires[False]
────────────────────────────────────────────────────────────────── WARNING ──────────────────────────────────────────────────────────────────
Could not convert the old property at key parameters with type SireMM::AmberParams to match the new, edited molecule. This property has been
deleted.
Error was SireError::incomplete_code: Cannot make compatible if atom order has changed!

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

────────────────────────────────────────────────────────────────── WARNING ──────────────────────────────────────────────────────────────────
Could not convert the old property at key parameters with type SireMM::AmberParams to match the new, edited molecule. This property has been
deleted.
Error was SireError::incomplete_code: Cannot make compatible if atom order has changed!

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squash_multires[True] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping[False] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping[True] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_atom_mapping_implicit[False] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_atom_mapping_implicit[True] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_atom_mapping_explicit[False] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_atom_mapping_explicit[True] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_unsquash[False]
────────────────────────────────────────────────────────────────── WARNING ──────────────────────────────────────────────────────────────────
Could not convert the old property at key parameters with type SireMM::AmberParams to match the new, edited molecule. This property has been
deleted.
Error was SireError::incomplete_code: Cannot make compatible if atom order has changed!

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

────────────────────────────────────────────────────────────────── WARNING ──────────────────────────────────────────────────────────────────
Could not convert the old property at key parameters with type SireMM::AmberParams to match the new, edited molecule. This property has been
deleted.
Error was SireError::incomplete_code: Cannot make compatible if atom order has changed!

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_unsquash[True] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_unsquash_multires[False]
────────────────────────────────────────────────────────────────── WARNING ──────────────────────────────────────────────────────────────────
Could not convert the old property at key parameters with type SireMM::AmberParams to match the new, edited molecule. This property has been
deleted.
Error was SireError::incomplete_code: Cannot make compatible if atom order has changed!

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

────────────────────────────────────────────────────────────────── WARNING ──────────────────────────────────────────────────────────────────
Could not convert the old property at key parameters with type SireMM::AmberParams to match the new, edited molecule. This property has been
deleted.
Error was SireError::incomplete_code: Cannot make compatible if atom order has changed!

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_unsquash_multires[True] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_decouple_lambda[decoupled_system-False-True-True] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_decouple_lambda[decoupled_system-True-True-False] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_decouple_lambda[redecoupled_system-False-True-False] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_decouple_lambda[redecoupled_system-True-True-True] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_decouple_lambda[decoupled_system-False-False-False] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_decouple_lambda[decoupled_system-True-False-False] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_decouple_lambda[redecoupled_system-False-False-False] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_decouple_lambda[redecoupled_system-True-False-False] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_order[decouple-merge-normal-order0] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_order[decouple-normal-merge-order1] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_order[merge-decouple-normal-order2] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_order[merge-normal-decouple-order3] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_order[normal-merge-decouple-order4] PASSED
tests/Sandpit/Exscientia/Align/test_squash.py::test_squashed_molecule_mapping_order[normal-decouple-merge-order5] PASSED

============================================================= warnings summary ==============================================================
```

**Comment by chryswoods** [comment]
Yes - these are properties that can't be automatically merged (and would have been silently dropped). There are a couple of solutions. One is to disable warnings while you are doing the merge (not great). Or the other is to remove the `AmberParams` object from the molecules before merging. I am thinking more generally to remove AmberParams because it duplicates information already in the rest of the molecule, and it is easy for it to get out of sync with the actual charges and other forcefield parameters.

**Comment by lohedges** [comment]
That makes sense. In this case it doesn't seem to happen during a merge, though. I can't quite figure out where exactly within the `squash` and `unsquash` functions this is occuring. A regular merge involving molecules containing the `parameters` property (which is an `AmberParams` object) causes no issues. Will keep digging.

**Comment by lohedges** [comment]
Okay, I've found where it's happening (within a private `_removeDummies` function). It's trivial to just remove the property before the partial molecule is extracted and it has no side effect in this instance. Will just do this.
